omnigibson/transition_rules.py

```diff
@@ -103,551 +263,952 @@ def __call__(self, obj):
+            filter_1_name (str): Name of the filter whose object candidates will be pruned based on whether or not
+                they are touching any object from @filter_2_name
+            filter_2_name (str): Name of the filter whose object candidates will be used to prune the candidates from
+                @filter_1_name
+        """
+        self._filter_1_name = filter_1_name
+        self._filter_2_name = filter_2_name
 
-                    {"apple": CategoryFilter("apple"), "knife": CategoryFilter("knife")},
+        # Will be filled in during self.initialize
+        # Maps object to the list of rigid body idxs in the global contact matrix corresponding to filter 1
+        self._filter_1_idxs = None
+        # List of rigid body idxs in the global contact matrix corresponding to filter 2
+        self._filter_2_idxs = None
 
-                the transition rule step will produce all 2-tuples of valid (apple, knife) combinations:
+    def refresh(self, object_candidates):
+        # Register idx mappings
+        self._filter_1_idxs = {obj: [RigidContactAPI.get_body_idx(link.prim_path) for link in obj.links.values()]
+                            for obj in object_candidates[self._filter_1_name]}
+        self._filter_2_idxs = [RigidContactAPI.get_body_idx(link.prim_path)
+                               for obj in object_candidates[self._filter_2_name] for link in obj.links.values()]
 
-                    {"apple": apple_i, "knife": knife_j}
+    def __call__(self, object_candidates):
+        # Get all impulses
+        impulses = RigidContactAPI.get_all_impulses()
 
-                based on the current instances of each object type in the scene and pass them to @self.condition as the
-                @individual_objects entry.
-                If None is specified, then no filter will be applied
+        # Keep any object that has non-zero impulses between itself and any of the @filter_2_name's objects
+        objs = []
+        for obj in object_candidates[self._filter_1_name]:
+            if np.any(impulses[self._filter_1_idxs[obj]][:, self._filter_2_idxs]):
+                objs.append(obj)
 
-            group_filters (None or dict): Group object filters that this filter cares about. For each name, filter
-                key-value pair, the global transition rule step will produce a single dictionary of valid filtered
-                objects.
-                For example, if the group filters are:
+        # Update candidates
+        object_candidates[self._filter_1_name] = objs
 
-                    {"apple": CategoryFilter("apple"), "knife": CategoryFilter("knife")},
+        # If objs is empty, return False, otherwise, True
+        return len(objs) > 0
 
-                the transition rule step will produce the following dictionary:
+    @classproperty
+    def modifies_filter_names(self):
+        # Only modifies values from filter 1
+        return set((self._filter_1_name,))
```

Comment:
why is this a set of tuple? Instead of a set?

Reply:
because `self._filter_1_name` is a str, so e.g.: `set("hello")` = `{'h', 'e', 'l', 'l', 'o'}` xD